server/src/main/java/org/apache/iotdb/db/mpp/execution/SchemaDriver.java

```diff
@@ -0,0 +1,146 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iotdb.db.mpp.execution;
+
+import org.apache.iotdb.db.mpp.buffer.ISinkHandle;
+import org.apache.iotdb.db.mpp.common.FragmentInstanceId;
+import org.apache.iotdb.db.mpp.operator.Operator;
+import org.apache.iotdb.tsfile.read.common.block.TsBlock;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import io.airlift.units.Duration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static org.apache.iotdb.db.mpp.operator.Operator.NOT_BLOCKED;
+
+@NotThreadSafe
+public class SchemaDriver implements ExecFragmentInstance {
+
+  private static final Logger logger = LoggerFactory.getLogger(DataDriver.class);
+
+  private final Operator root;
+  private final ISinkHandle sinkHandle;
+  private final SchemaDriverContext driverContext;
+
+  private final AtomicReference<SettableFuture<Void>> driverBlockedFuture = new AtomicReference<>();
+
+  public SchemaDriver(Operator root, ISinkHandle sinkHandle, SchemaDriverContext driverContext) {
+    this.root = root;
+    this.sinkHandle = sinkHandle;
+    this.driverContext = driverContext;
+    // initially the driverBlockedFuture is not blocked (it is completed)
+    SettableFuture<Void> future = SettableFuture.create();
+    future.set(null);
+    driverBlockedFuture.set(future);
+  }
+
+  @Override
+  public boolean isFinished() {
+    try {
+      boolean isFinished = driverBlockedFuture.get().isDone() && root != null && root.isFinished();
+      if (isFinished) {
+        driverContext.finish();
+      }
+      return isFinished;
+    } catch (Throwable t) {
+      logger.error(
+          "Failed to query whether the schema driver {} is finished", driverContext.getId(), t);
+      driverContext.failed(t);
+      return true;
+    }
+  }
+
+  @Override
+  public ListenableFuture<Void> processFor(Duration duration) {
+    // if the driver is blocked we don't need to continue
+    SettableFuture<Void> blockedFuture = driverBlockedFuture.get();
+    if (!blockedFuture.isDone()) {
+      return blockedFuture;
+    }
+
+    long maxRuntime = duration.roundTo(TimeUnit.NANOSECONDS);
+
+    long start = System.nanoTime();
+    try {
+      do {
+        ListenableFuture<Void> future = processInternal();
+        if (!future.isDone()) {
+          return updateDriverBlockedFuture(future);
+        }
+      } while (System.nanoTime() - start < maxRuntime && !root.isFinished());
+    } catch (Throwable t) {
+      logger.error("Failed to execute fragment instance {}", driverContext.getId(), t);
+      driverContext.failed(t);
+      close();
+    }
+    return NOT_BLOCKED;
```

Comment:
The same as `DataDriver`. It should return an `isCancelled() == true` `ListenableFuture` to tell the scheduler that it's stopped abnormally.

Reply:
fixed